### PR TITLE
Runtime: more efficient (space-wise) marshalling.

### PR DESCRIPTION
**Pull request by corwin-of-amber**
Modified caml_output_val to correctly reflects sharing.
As a result, it avoids duplication and saves a lot of space.

This has been thoroughly tested by compiling Coq `.vo` files, where it can reduce the file size by an order of magnitude, and produces roughly the same size as the native `coqc`.

### CODE REVIEW

**Comment by hhugo** [comment]
Fixes https://github.com/ocsigen/js_of_ocaml/issues/359

**Comment by vouillon** [comment]
What about using a `WeakMap`?

**Comment by corwin-of-amber** [comment]
> What about using a `WeakMap`?

Or a regular `Map` for that matter... yes, that could be done and will improve the time complexity (saving the need for linear search in the internal object table). Question is if it's ok to rely on ES2015 specs in the runtime portion.

**Comment by hhugo** [comment]
Thanks a lot for this.

Would you mind writing some tests ? There are plenty of examples in https://github.com/ocsigen/js_of_ocaml/blob/master/compiler/tests/.

It would be nice to demonstrate that cyclic value can now be marshaled.

**Comment by corwin-of-amber** [comment]
Alas, I failed to build the latest version from source, due to 
`Error: Unbound value Ast_mapper.make_error_of_message`

https://github.com/ocsigen/js_of_ocaml/blob/8022d1eafe263bef4c6cb4218246ef1709ee55da/ppx/ppx_js/lib_internal/ppx_js_internal.ml#L462

It seemed to work fine in Travis, so I looked at the configuration file and was not sure how to proceed, does the script (`tools/travis.sh`) even compile the source?

**Comment by corwin-of-amber** [comment]
Ok I ran `opam install -y cohttp-lwt-unix menhir ppx_expect` and now it builds except for the missing `graphics` package.

(I think it caused it to upgrade `ocaml-migrate-parsetree`, which for some reason, was not done by `opam install -y --deps-only js_of_ocaml`.)

**Comment by hhugo** [comment]
The `ocaml-migrate-parsetree {>= "1.3.1"}` constraint is not knonw to the opam-repository yet.
I would not expect opam to upgrade unless you pin js_of_ocaml.

About the missing graphics lib, I think building the compiler and its (new) tests should be enough (`dune build @compiler/runtest`)

**Comment by corwin-of-amber** [comment]
Added one test for shared data (DAG) and one for cycles (based on #359).

I noticed that for the other tests in `channel.ml`, you added OCaml counterparts. Do you want that for the two new ones as well?

**Comment by hhugo** [comment]
Tests look good as they are now. Maybe move them in their own file.

**Comment by corwin-of-amber** [comment]
Ok, suggestions on how to name it? I can't use `compiler/tests/marshal.ml` because it clashes with stdlib.

**Comment by hhugo** [comment]
To follow on @vouillon remark. Could we you `weakMap/Map` with some fallback if not available ?

We should probably also check the flags in `caml_output_value`
https://github.com/ocsigen/js_of_ocaml/blob/master/runtime/io.js#L389 and 
`caml_output_value_to_{buffer,string,bytes}` https://github.com/ocsigen/js_of_ocaml/blob/master/runtime/marshal.js#L445

**Comment by corwin-of-amber** [comment]
> To follow on @vouillon remark. Could we you `weakMap/Map` with some fallback if not available ?

Good idea.

> We should probably also check the flags in `caml_output_value`

Can easily support `Marshal.No_sharing` and `Marshal.Compat32`, I don't know about `Marshal.Closures`.
Are the flags the same for `caml_output_value` and for `caml_output_val`?

File: runtime/marshal.js [hunk 1]
```diff
@@ -337,10 +337,46 @@ function caml_marshal_data_size (s, ofs) {
   return (get32(s, ofs + 4));
 }
 
+//Provides: MlObjectTable
+var MlObjectTable;
+if (typeof joo_global_object.WeakMap === 'undefined') {
+  MlObjectTable = function() {
```
Comment (corwin-of-amber) [inline runtime/marshal.js]: In the current test setup this code is never tested because the condition is always `false`. 😞

**Comment by corwin-of-amber** [comment]
Ok, flags are in.

File: runtime/marshal.js [hunk 2]
```diff
@@ -365,13 +406,31 @@ var caml_output_val = function (){
...
+      joo_global_object.console.warn("in caml_output_val: flag Marshal.Closures is not supported.");
+
     var writer = new Writer ();
     var stack = [];
+    var intern_obj_table = new MlObjectTable();
+
+    function memo(v) {
+      var existing_offset = no_sharing ? undefined : intern_obj_table.recall(v);
+      if (existing_offset) { writer.write_shared(existing_offset); return existing_offset; }
+      else intern_obj_table.store(v);
```
Comment (hhugo) [inline runtime/marshal.js]: do we need to call `intern_obj_table.store` when `no_sharing` is defined ?

File: runtime/marshal.js [hunk 3]
```diff
@@ -337,10 +337,46 @@ function caml_marshal_data_size (s, ofs) {
...
+if (typeof joo_global_object.WeakMap === 'undefined') {
+  MlObjectTable = function() {
+    /* polyfill */
+    function NaiveLookup(objs) { this.objs = objs; }
+    NaiveLookup.prototype.get = function(v) {
+      for (var i = 0; i < this.objs.length; i++) {
+        if (this.objs[i] === v) return i;
+      }
+    };
+    NaiveLookup.prototype.set = function() { };
```
Comment (hhugo) [inline runtime/marshal.js]: Why is `set` empty ?

Comment (hhugo) [reply runtime/marshal.js]: I understand now.
This probably deserves a small comment, saying that `objs` is shared between `MlObjectTable` and `NaiveLookup` and that `store` is responsible for updating `objs`

File: compiler/tests/std_marshal.ml [hunk 4]
```diff
@@ -0,0 +1,82 @@
+(* Js_of_ocaml tests
+ * http://www.ocsigen.org/js_of_ocaml/
+ * Copyright (C) 2019 Hugo Heuzard
```
Comment (hhugo) [inline compiler/tests/std_marshal.ml]: Update the name here

Comment (corwin-of-amber) [reply compiler/tests/std_marshal.ml]: Ah oops.

File: runtime/marshal.js (hunk 2 above)
Comment (corwin-of-amber) [reply runtime/marshal.js]: We still need to update the object counter.
I can do `writer.obj_counter++` instead, and remove the assignment `writer.obj_counter = intern_obj_table.length` at the end of the function.

Comment (hhugo) [reply runtime/marshal.js]: obj_counter seems to only be used when sharing is enable.
https://github.com/ocaml/ocaml/blob/trunk/runtime/extern.c#L180

File: runtime/marshal.js (hunk 1 above)
Comment (hhugo) [reply runtime/marshal.js]: I've tested the code manually.

**Comment by hhugo** [comment]
I took the liberty to push to your branch directly (hope you don't mind).
Could you review my changes ?

**Comment by corwin-of-amber** [comment]
Looking great.
I noticed that `memo(v)` can in fact return a boolean. So I tweaked it further.

**Comment by hhugo** [comment]
Thanks a lot !

**Pull request merged**